In [1]:
import pandas as pd

In [2]:
df_survey = pd.read_csv('survey_results_public.csv', low_memory=False) # df_survey - сирі перші дані
df_schema = pd.read_csv('survey_results_schema.csv') # df_schema - сирі другі дані

Завдання 1 — Підрахунок респондентів в опитуванні. Спочатку подивимось на всі колонки df_survey. Для цього змінемо налаштування Рandas (pd.set_option) та назвемо конкретно, що хочемо налаштувати ('display.max_columns', None), де None - показати без обмежень всі колонки.

In [3]:
pd.set_option('display.max_columns', None)
df_survey.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,AILearnHow,YearsCode,DevType,OrgSize,ICorPM,RemoteWork,PurchaseInfluence,TechEndorseIntro,TechEndorse_1,TechEndorse_2,TechEndorse_3,TechEndorse_4,TechEndorse_5,TechEndorse_6,TechEndorse_7,TechEndorse_8,TechEndorse_9,TechEndorse_13,TechEndorse_13_TEXT,TechOppose_1,TechOppose_2,TechOppose_3,TechOppose_5,TechOppose_7,TechOppose_9,TechOppose_11,TechOppose_13,TechOppose_16,TechOppose_15,TechOppose_15_TEXT,Industry,JobSatPoints_1,JobSatPoints_2,JobSatPoints_3,JobSatPoints_4,JobSatPoints_5,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,JobSatPoints_13,JobSatPoints_14,JobSatPoints_15,JobSatPoints_16,JobSatPoints_15_TEXT,AIThreat,NewRole,ToolCountWork,ToolCountPersonal,Country,Currency,CompTotal,LanguageChoice,LanguageHaveWorkedWith,LanguageWantToWorkWith,LanguageAdmired,LanguagesHaveEntry,LanguagesWantEntry,DatabaseChoice,DatabaseHaveWorkedWith,DatabaseWantToWorkWith,DatabaseAdmired,DatabaseHaveEntry,DatabaseWantEntry,PlatformChoice,PlatformHaveWorkedWith,PlatformWantToWorkWith,PlatformAdmired,PlatformHaveEntry,PlatformWantEntry,WebframeChoice,WebframeHaveWorkedWith,WebframeWantToWorkWith,WebframeAdmired,WebframeHaveEntry,WebframeWantEntry,DevEnvsChoice,DevEnvsHaveWorkedWith,DevEnvsWantToWorkWith,DevEnvsAdmired,DevEnvHaveEntry,DevEnvWantEntry,SOTagsHaveWorkedWith,SOTagsWantToWorkWith,SOTagsAdmired,SOTagsHaveEntry,SOTagsWant Entry,OpSysPersonal use,OpSysProfessional use,OfficeStackAsyncHaveWorkedWith,OfficeStackAsyncWantToWorkWith,OfficeStackAsyncAdmired,OfficeStackHaveEntry,OfficeStackWantEntry,CommPlatformHaveWorkedWith,CommPlatformWantToWorkWith,CommPlatformAdmired,CommPlatformHaveEntr,CommPlatformWantEntr,AIModelsChoice,AIModelsHaveWorkedWith,AIModelsWantToWorkWith,AIModelsAdmired,AIModelsHaveEntry,AIModelsWantEntry,SOAccount,SOVisitFreq,SODuration,SOPartFreq,SO_Dev_Content,SO_Actions_1,SO_Actions_16,SO_Actions_3,SO_Actions_4,SO_Actions_5,SO_Actions_6,SO_Actions_9,SO_Actions_7,SO_Actions_10,SO_Actions_15,SO_Actions_15_TEXT,SOComm,SOFriction,AISelect,AISent,AIAcc,AIComplex,AIToolCurrently partially AI,AIToolDon't plan to use AI for this task,AIToolPlan to partially use AI,AIToolPlan to mostly use AI,AIToolCurrently mostly AI,AIFrustration,AIExplain,AIAgents,AIAgentChange,AIAgent_Uses,AgentUsesGeneral,AIAgentImpactSomewhat agree,AIAgentImpactNeutral,AIAgentImpactSomewhat disagree,AIAgentImpactStrongly agree,AIAgentImpactStrongly disagree,AIAgentChallengesNeutral,AIAgentChallengesSomewhat disagree,AIAgentChallengesStrongly agree,AIAgentChallengesSomewhat agree,AIAgentChallengesStrongly disagree,AIAgentKnowledge,AIAgentKnowWrite,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",AI CodeGen tools or AI-enabled apps,14.0,"Developer, mobile",20 to 99 employees,People manager,Remote,"Yes, I influenced the purchase of a substantia...",Work,10.0,7.0,9.0,6.0,3.0,11.0,12.0,1.0,8.0,14.0,NaN,15.0,7.0,8.0,12.0,11.0,1.0,6.0,13.0,3.0,16.0,NaN,Fintech,3.0,1.0,4.0,9.0,5.0,10.0,12.0,11.0,2.0,6.0,7.0,13.0,14.0,15.0,8.0,NaN,I'm not sure,I have neither consider or transitioned into a...,7.0,3.0,Ukraine,EUR European Euro,52800.0,Yes,Bash/Shell (all shells);Dart;SQL,Dart,Dart,NaN,NaN,Yes,Cloud Firestore;PostgreSQL,NaN,NaN,NaN,NaN,Yes,Amazon Web Services (AWS);Cloudflare;Firebase;...,NaN,NaN,NaN,NaN,No,NaN,NaN,NaN,NaN,NaN,Yes,Android Studio;Notepad++;Visual Studio;Visual ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Windows;MacOS;Android,Windows;MacOS;Android;iOS;iPadOS,Confluence;GitHub;GitLab;Jira;Markdo

Рахуємо загальну кількість респондентів через .nunique() -  number of unique values = кількість унікальних значень.

In [4]:
total_respondents = df_survey['ResponseId'].nunique()
print(f"Загальна кількість респондентів: {total_respondents}")

Загальна кількість респондентів: 49191


Завдання 2. 
Аналіз повноти відповідей респондентів. 
Спочатку подивимось на список колонок і початок таблиці df_schema.
І тут ми бачимо, що колонка qname містить самі назви питань (наприклад TechEndorse_1, TechEndorse_2 і т.д.) — саме їх треба порівняти з колонками у основному датасеті.

In [5]:
print(df_schema.columns.tolist())
print(df_schema.head())          

['qid', 'qname', 'question', 'type', 'sub', 'sq_id']
     qid          qname                                           question  \
0  QID18  TechEndorse_1  What attracts you to a technology or causes yo...   
1  QID18  TechEndorse_2  What attracts you to a technology or causes yo...   
2  QID18  TechEndorse_3  What attracts you to a technology or causes yo...   
3  QID18  TechEndorse_4  What attracts you to a technology or causes yo...   
4  QID18  TechEndorse_5  What attracts you to a technology or causes yo...   

  type                                      sub  sq_id  
0   RO  AI integration or AI Agent capabilities    1.0  
1   RO                          Easy-to-use API    2.0  
2   RO                  Robust and complete API    3.0  
3   RO     Customizable and manageable codebase    4.0  
4   RO                   Reputation for quality    5.0  


1. Робимо нову множину schema_questions - беремо колонку qname з df_schema - df_schema['qname'] - та перетврюємо колонку спочатку в список  .tolist(), а потім в множину - колекцію унікальних значень без повторів за допомогою - set(...), бо множини вміють робити перетин (інтерсекцію).
2. Виводимо за допомогою print та f"..." (це f-рядок (f-string) - спосіб вставити змінну прямо в текст) кількість питань у  множині schema_questions.
3. Робимо нову множину survey_columns з df_survey таким же чином, через columns.tolist() та set(...).
4. Знаходимо перетин - це знов нова множина common_columns, за допомогою intersection.
5. Виводимо за допомогою print та f"..." кількість спільних колонок множини common_columns.

In [6]:
schema_questions = set(df_schema['qname'].tolist())
print(f"Кількість питань у схемі: {len(schema_questions)}")

survey_columns = set(df_survey.columns.tolist())
common_columns = schema_questions.intersection(survey_columns) 
print(f"Кількість спільних колонок: {len(common_columns)}")

Кількість питань у схемі: 139
Кількість спільних колонок: 126


Тепер можна підрахувати кількість респондентів - complete_responses (повна відповідь), які відповіли на всі питання. Для цього: 
1. перетворюємо множину в список: list(common_columns),
2. беремо тільки 126 колонок з питаннями df_survey[list(common_columns)],
3. видаляємо всі рядки де є хоча б один пропуск (NaN) за допомогою .dropna(),
4. рахуємо скільки рядків залишилось за допомогою .shape[0]

In [7]:
complete_responses = df_survey[list(common_columns)].dropna().shape[0] # shape[0] - повертає рядки, бо [1] - це колонки
print(f"Кількість респондентів без пропущених значень: {complete_responses}")

Кількість респондентів без пропущених значень: 0


1. Така відповідь не показує повну картину, тому, щоб краще зрозуміти дані, треба побачити, наскільки неповними є відповіді в середньому. Для цього знаходимо один стовпчик, який нам покаже, скільки питань пропущено на одну людину - називаємо його missing_per_person (це тип даних Series), беремо тільки колонки з питаннями df_survey[list(common_columns)], перевіряємо кожну клітинку — чи є там пропуск? (Повертає True якщо пропуск, False якщо є значення) за допомогою .isnull(), рахуємо суму по кожному рядку  .sum(axis=1), де axis=1 - йде вздовж рядка; True = 1, False = 0 — тому сума = кількість пропусків.
2. Виводимо за допомогою print та f-рядка min, median та max.

In [8]:
missing_per_person = df_survey[list(common_columns)].isnull().sum(axis=1)
print(f"Мінімум пропущених питань на людину: {missing_per_person.min()}")
print(f"Медіана пропущених питань на людину: {missing_per_person.median()}")
print(f"Максимум пропущених питань на людину: {missing_per_person.max()}")

Мінімум пропущених питань на людину: 4
Медіана пропущених питань на людину: 50.0
Максимум пропущених питань на людину: 124


Висновок по завданню 2:
Кількість респондентів без пропущених значень дорівнює 0 - це правильна відповідь. Бо навіть самі активні респонденти пропустили мінімум 4 питання. Типова людина, що стоїть рівно посередині інших людей (де половина пропустила більше питань, а половина - менше), пропустила до 50 питань. А хтось взагалі відповів тільки на два питання зі 126. 

Завдання 3. Статистичний аналіз досвіду респондентів. Для цього глянемо на колонку WorkExp. І тут ми бачимо, що колонка WorkExp містить числа (роки досвіду), її тип - float64, це дробові числа. Розрахунки робити можна.

In [9]:
print(df_survey['WorkExp'].dtype)
print(df_survey['WorkExp'].head(10))

float64
0     8.0
1     2.0
2    10.0
3     4.0
4    21.0
5    15.0
6     9.0
7    22.0
8     9.0
9     7.0
Name: WorkExp, dtype: float64


Знаходимо міри центральної тенденції (пропущені значення (NaN) всі три методи ігнорують автоматично). Це середнє значення (mean), медіана (median) та мода (mode). (Значень моди може бути кілька, .mode() повертає Series (список мод), тому за допомогою [0] ми скажемо - взяти перше значення.)

In [10]:
mean_exp = df_survey['WorkExp'].mean()
median_exp = df_survey['WorkExp'].median()
mode_exp = df_survey['WorkExp'].mode()[0] 

print(f"Середнє значення досвіду: {mean_exp:.1f} років") # :.1f — виводимо з одним знаком після коми
print(f"Медіана досвіду:        {median_exp:.1f} років")
print(f"Мода досвіду:          {mode_exp:.1f} років")

Середнє значення досвіду: 13.4 років
Медіана досвіду:        10.0 років
Мода досвіду:          10.0 років


Висновок:

1. Середнє 13.4 — це коли весь досвід порівну розділити між усіма респондентами.
2. Медіана 10.0 — це коли половина розробників має менше 10 років досвіду, половина більше.
3. Мода 10.0 — відповідь, яка зустрічається найчастіше - 10 років.

Середнє більше за медіану — це означає що є люди і їх небагато, у яких досвід дуже великий, більше на 20 чи 30 років, ніж у інших в 10 років. Тому середнє буде вище за медіану. Медіана в таких випадках точніше відображає "типового" розробника.

Завдання 4. Аналіз віддаленої роботи. 
 1. Спочатку треба знайти колонку, що відповідає за віддалену роботу. Для цього шукаємо за словом у назві, за допомогою df_survey.columns — список всіх назв колонок;
 2. for col in df_survey.columns — перебираємо кожну назву по одній;
 3. .lower() — перетворюємо назву в маленькі літери, щоб не пропустити якщо написано Remote або REMOTE або remote;
 4. if 'remote' in col.lower() — перевіряємо чи є слово remote в назві;
 5. or 'work' in col.lower() — або чи є слово work.
 6. В кінці генеруємо список remote_columns з цими словами - 'remote' або 'work'.

In [11]:
remote_columns = [col for col in df_survey.columns if 'remote' in col.lower() or 'work' in col.lower()]
print(remote_columns)

['WorkExp', 'RemoteWork', 'ToolCountWork', 'LanguageHaveWorkedWith', 'LanguageWantToWorkWith', 'DatabaseHaveWorkedWith', 'DatabaseWantToWorkWith', 'PlatformHaveWorkedWith', 'PlatformWantToWorkWith', 'WebframeHaveWorkedWith', 'WebframeWantToWorkWith', 'DevEnvsHaveWorkedWith', 'DevEnvsWantToWorkWith', 'SOTagsHaveWorkedWith', 'SOTagsWantToWorkWith', 'OfficeStackAsyncHaveWorkedWith', 'OfficeStackAsyncWantToWorkWith', 'CommPlatformHaveWorkedWith', 'CommPlatformWantToWorkWith', 'AIModelsHaveWorkedWith', 'AIModelsWantToWorkWith']


І як бачимо, ця колонка називається 'RemoteWork'. Дивимось які значення там є за допомогою value_counts() яке поверне кількість унікальних збігів у певній колонці 

In [12]:
print(df_survey['RemoteWork'].value_counts())

Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: RemoteWork, dtype: int64


І знайшли. Повністю віддалено — це відповідь - Remote. 
Тепер треба знайти всіх хто має Remote в колонці RemoteWork і порахувати їх. 
1. Для цього создамо - remote_count. Та візьмемо один стовпчик з таблиці df_survey['RemoteWork'].
2. Перевіремо кожен рядок, чи дорівнює він відповіді 'Remote'? df_survey['RemoteWork'] == 'Remote'
3. Фільтруємо таблицю, в ній залишаться тільки рядки, де відповідь - True
4. Рахуємо, скільки рядків залишилось - .shape[0] (повертає два числа у вигляді кортежу, [0] - кількість рядків, [1] - кількість колонок)

In [13]:
remote_count = df_survey[df_survey['RemoteWork'] == 'Remote'].shape[0]
print(f"Кількість респондентів, які працюють віддалено: {remote_count}")

Кількість респондентів, які працюють віддалено: 10931


Висновок: повністю віддалено працює 10 931 респондент — це приблизно 22% від загальної кількості (49 191).

Завдання 5. Визначення популярності Python. Спочатку треба знайти необхідну колонку зі словом 'language', через список lang_columns. 
1. Тут df_survey.columns — список всіх назв колонок.
2. for col in df_survey.columns — перебираємо кожну назву по одній.
3. col.lower() — перетворюємо назву в маленькі літери.
4. if 'language' in col.lower() — перевіряємо чи є слово language в назві.
5. [col for col in ...] — збираємо тільки ті що пройшли перевірку.
6. Хоча у завданні 4 ми вже побачили цю потрібну нам колонку - 'LanguageHaveWorkedWith', тому тільки перевіряємо.

In [14]:
lang_columns = [col for col in df_survey.columns if 'language' in col.lower()]
print(lang_columns)

['LanguageChoice', 'LanguageHaveWorkedWith', 'LanguageWantToWorkWith', 'LanguageAdmired', 'LanguagesHaveEntry', 'LanguagesWantEntry']


Дивимось перші значення колонки 'LanguageHaveWorkedWith' (мова, якою працювали по програмуванню).

In [15]:
print(df_survey['LanguageHaveWorkedWith'].head(10))

0                     Bash/Shell (all shells);Dart;SQL
1                                                 Java
2                  Dart;HTML/CSS;JavaScript;TypeScript
3                                      Java;Kotlin;SQL
4    C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;P...
5                                           Java;Scala
6                                JavaScript;TypeScript
7    Bash/Shell (all shells);HTML/CSS;JavaScript;Py...
8                                    Java;Python;Scala
9                                                  NaN
Name: LanguageHaveWorkedWith, dtype: object


Ми бачимо Python та інші мови, які розділені крапкою з комою.
 
 1. Создаємо python_count та шукаємо слово 'Python' в кожному рядку за допомогою  .str.contains('Python', na=False), na=False означає, що пропуски рахуємо як "не містить" замість помилки.
 2. Рахуємо кількість True (тобто тих хто має Python) за допомогою  .sum()
 3. Знаходимо відсоток респондентів, які програмують на Python через (python_count / total_respondents) * 100.
 4. .1f - це одне число після коми

In [16]:
python_count = df_survey['LanguageHaveWorkedWith'].str.contains('Python', na=False).sum()
python_percent = (python_count / total_respondents) * 100

print(f"Кількість респондентів, які програмують на Python: {python_count}")
print(f"Відсоток респондентів, які програмують на Python: {python_percent:.1f}%")

Кількість респондентів, які програмують на Python: 18466
Відсоток респондентів, які програмують на Python: 37.5%


Висновок: 37.5% респондентів програмують на Python, це майже кожен третій респондент. Ця мова дуже популярна.

Завдання 6. Аналіз шляхів навчання програмуванню. Спочатку шукаємо потрібну колонку за допомогою слова learn і дивимось які значення там є.

In [17]:
learn_columns = [col for col in df_survey.columns if 'learn' in col.lower()]
print(learn_columns)

['LearnCodeChoose', 'LearnCode', 'LearnCodeAI', 'AILearnHow']


Тепер дивимось що міститься в колонці LearnCode:

In [18]:
print(df_survey['LearnCode'].head(10))

0    Online Courses or Certification (includes all ...
1    Online Courses or Certification (includes all ...
2    Online Courses or Certification (includes all ...
3    Other online resources (e.g. standard search, ...
4                                                  NaN
5    Online Courses or Certification (includes all ...
6    Videos (not associated with specific online co...
7    Books / Physical media;Technical documentation...
8    Online Courses or Certification (includes all ...
9    Other online resources (e.g. standard search, ...
Name: LearnCode, dtype: object


Бачимо що онлайн курси називаються Online Courses or Certification та є інші курси, всі розділені крапкою з комою.
 Создаємо online_courses_count та шукаємо в 'LearnCode' - 'Online Courses or Certification' в кожному рядку за допомогою  .str.contains('Online Courses or Certification', na=False), na=False означає, що пропуски рахуємо як "не містить" замість помилки. 
 Рахуємо кількість True (тобто тих хто має 'Online Courses or Certification') за допомогою .sum()

In [19]:
online_courses_count = df_survey['LearnCode'].str.contains('Online Courses or Certification', na=False).sum() 
print(f"Кількість респондентів, які навчались через онлайн курси: {online_courses_count}")

Кількість респондентів, які навчались через онлайн курси: 10973


Завдання 7. Географічний аналіз компенсації Python-розробників. Спочатку перевіримо колонку з компенсацією ConvertedCompYearly (річна компенсація в долларах).

In [20]:
print(df_survey['ConvertedCompYearly'].head(10))

0     61256.0
1    104413.0
2     53061.0
3     36197.0
4     60000.0
5    120000.0
6      6240.0
7     72000.0
8     70000.0
9      5400.0
Name: ConvertedCompYearly, dtype: float64


 1. Фільтруємо тільки Python розробників. Создаємо python_devs — датасет тільки з розробниками на Python, та шукаємо в колонці 'LanguageHaveWorkedWith', в кожному рядку відповідь 'Python' за допомогою  .str.contains('Python', na=False) - .str.contains() — це пошук тексту всередині рядка, .str — я хочу працювати з цим як з текстом.
 2. , na=False означає, що пропуски рахуємо як "не містить" замість помилки. Таблиця фільтрується і в ній залишаються тільки рядки, де відповідь - True.
 3. Создаємо таблицю compensation_by_country та групуємо по країнах - .groupby('Country'), .agg() — одночасно рахуємо і середнє і медіану, .round(0) — округляємо до цілих чисел.
 4. Сортуємо за середньою компенсацією в таблиці compensation_by_country і виводимо результат.

In [21]:
python_devs = df_survey[df_survey['LanguageHaveWorkedWith'].str.contains('Python', na=False)] 
compensation_by_country = python_devs.groupby('Country')['ConvertedCompYearly'].agg(
    Середнє = 'mean',
    Медіана = 'median'
).round(0)

compensation_by_country = compensation_by_country.sort_values('Середнє', ascending=False)
print(compensation_by_country.head(20))

                                                     Середнє   Медіана
Country                                                               
Oman                                                390135.0  390135.0
Andorra                                             226104.0  226104.0
Viet Nam                                            218837.0    8254.0
United States of America                            173299.0  150000.0
Switzerland                                         156457.0  142592.0
Singapore                                           147515.0   91391.0
Israel                                              135828.0  142594.0
Nigeria                                             128773.0    1808.0
Ireland                                             120524.0  116015.0
Nomadic                                             120132.0  139218.0
Australia                                           118091.0   97514.0
United Kingdom of Great Britain and Northern Ir...  117662.0   94958.0
Luxemb

Висновок:
1. Оман по компенсації на першому місці, Андора після нього, але середнє дорівнює медіані і це означає що там лише один Python розробник у датасеті, тому не показово.
2. США — найбільша вибірка з високою компенсацією: середня $173k, медіана $150k.
3. В'єтнам і Нігерія — великий розрив між середнім і медіаною — кілька дуже багатих респондентів "тягнуть" середнє вгору.
4. Швейцарія — стабільно висока компенсація, середня і медіана близькі.
5. Ізраіль - теж більш-менш стабільно висока компенсація, медіана вище за середнє, це означає більшість типових респондентів, та багатих респондентів немає. 

Завдання 8. Аналіз освіти найбільш оплачуваних спеціалістів. 
1. .nlargest(5, 'ConvertedCompYearly') — знаходимо 5 найбільших значень в колонці ConvertedCompYearly.
2. [['ConvertedCompYearly', 'EdLevel']] — показуємо тільки дві колонки, подвійні дужки та дві колонки означають, що це DataFrame (таблиця), навіть якщо всередині була одна колонка, це завжди DataFrame (одинарні дужки → повертає Series - один стовпчик).

In [22]:
top5 = df_survey.nlargest(5, 'ConvertedCompYearly')[['ConvertedCompYearly', 'EdLevel']]
print(top5)

       ConvertedCompYearly                                          EdLevel
34267           50000000.0              Associate degree (A.A., A.S., etc.)
28700           33552715.0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)
43143           18387548.0              Associate degree (A.A., A.S., etc.)
35353           15430267.0     Bachelor’s degree (B.A., B.S., B.Eng., etc.)
45971           13921760.0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)


Що бачимо:

Топ-5 найбільш оплачуваних заробляють від $13.9M  до  $50M на рік — це скоріш за все аномалії або помилки у даних, бо такі суми нереалістичні для найманих працівників. 

Освіта: 2 з Associate degree, 2 з Master's degree, 1 з Bachelor's degree. 

А найвищу компенсацію має людина з молодшим ступенем (Associate degree), не докторантурою.

*Завдання 9. Аналіз популярності Python по віковим категоріям.
1. Спочатку дивимось унікальні вікові категорії. df_survey['Age'] — беремо один стовпчик з таблиці.
2. .value_counts() — рахуємо, скільки разів зустрічається кожне значення і сортуємо від найбільшого до найменшого.

In [23]:
print(df_survey['Age'].value_counts())

25-34 years old      16519
35-44 years old      13241
18-24 years old       9210
45-54 years old       6275
55-64 years old       2626
65 years or older      942
Prefer not to say      378
Name: Age, dtype: int64


1. Створюємо нову колонку в df_survey, називаємо її 'UsesPython' (True/False) - чи використовує людина Python.
2. .str - працюємо з колонкою 'LanguageHaveWorkedWith', як з текстом.
3. Перевіряємо .contains — кожну клітинку — чи є всередині потрібне слово 'Python', щоб  не  було помилки - na=False.
4. Розкладаємо людей по групах за віком python_by_age. Групуємо по віку та по користуванням Python.
5. .agg - одночасно рахуємо - Всього='count', Python_розробників='sum'.
6. Для кожної групи рахуємо всього людей і скільки з них Python розробники.
7. .agg - одночасно рахуємо - рбимо нові стовпчики - Всього ='count', Python_розробників ='sum'.
8. Рахуємо відсоток Python розробників в кожній групі. Для цього добавляємо новий стовпчик 'Відсоток_Python' до датасету python_by_age через python_by_age['Python_розробників'] / python_by_age['Всього'] * 100. (.round(1) - округлюємо до 1 знака
9. Сортуємо від найбільшого відсотку до найменшого.
10. Виводимо результат.

In [24]:
df_survey['UsesPython'] = df_survey['LanguageHaveWorkedWith'].str.contains('Python', na=False)
python_by_age = df_survey.groupby('Age')['UsesPython'].agg(
    Всього ='count',
    Python_розробників ='sum'
)

python_by_age['Відсоток_Python'] = (python_by_age['Python_розробників'] / python_by_age['Всього'] * 100).round(1)
python_by_age = python_by_age.sort_values('Відсоток_Python', ascending=False)
print(python_by_age)

                   Всього  Python_розробників  Відсоток_Python
Age                                                           
18-24 years old      9210                3684             40.0
45-54 years old      6275                2424             38.6
55-64 years old      2626                 978             37.2
25-34 years old     16519                6102             36.9
35-44 years old     13241                4862             36.7
65 years or older     942                 298             31.6
Prefer not to say     378                 118             31.2


Цікаві спостереження:

Наймолодші (18-24) використовують Python найбільше — 40% — мабуть тому що Python зараз активно викладають у навчальних закладах. 

Найстарші (65+) використовують Python найменше — 31.6% — вони навчались програмувати коли Python ще не був таким популярним. 

Різниця між віковими групами відносно невелика (31-40%) — Python популярний у всіх вікових групах

Що ми бачимо:
1. Найвищий відсоток використання Python у наймолодшій групі - 18-24 років - 40.0 %. Це мабудь тому, що вони получають найсучасні знання.
2. Люди з найстарішим віком - від 65 років і вище - найменше використовуют Python - 31.6 %.
3. Але в основному, різниця в використвуванні Python невелика - від 31.2 до 40.0 %. І це може свідчити про його популярність. 

*Завдання 10. Аналіз індустрій серед високооплачуваних віддалених працівників. Спочатку знайдемо колонку з індустріями - industry_columns.

In [25]:
industry_columns = [col for col in df_survey.columns if 'industry' in col.lower() or 'sector' in col.lower()]
print(industry_columns)

['Industry']


Колонка називається Industry.
1. Знаходимо респондентів, які знаходяться у 75 перцентилі за компенсацією (тобто заробляють більше, ніж 75% респондентів). Рахуємо 75-й перцентиль компенсації та виводимо на екран через  f-рядок, де :,.0f - форматуємо число, : — означає "починається форматування", , — розділяє тисячі комою, .0 — кількість знаків після крапки - в нас 0f.  
2. Фільтруємо за двома критеріями - висока компенсація І віддалена робота - high_paid_remote.
3. Знаходимо найпопулярніші індустрії - .value_counts() — рахує частоту кожної індустрії і сортує за спаданням.

In [26]:
percentile_75 = df_survey['ConvertedCompYearly'].quantile(0.75)
print(f"75-й перцентиль компенсації: ${percentile_75:,.0f}")

high_paid_remote = df_survey[
    (df_survey['ConvertedCompYearly'] > percentile_75) & # & — оператор "І" для двох умов одночасно
    (df_survey['RemoteWork'] == 'Remote')
]
print(f"Кількість відфільтрованих респондентів: {len(high_paid_remote)}")

top_industries = high_paid_remote['Industry'].value_counts().head(10)

print("\nТоп-10 індустрій серед високооплачуваних віддалених працівників:")
print(top_industries)

75-й перцентиль компенсації: $120,596
Кількість відфільтрованих респондентів: 2461

Топ-10 індустрій серед високооплачуваних віддалених працівників:
Software Development                          1186
Fintech                                        190
Healthcare                                     188
Other:                                         176
Internet, Telecomm or Information Services     138
Banking/Financial Services                      88
Government                                      78
Media & Advertising Services                    75
Retail and Consumer Services                    65
Transportation, or Supply Chain                 63
Name: Industry, dtype: int64


Висновок: 
1. На першому місці, з великим відривом, серед високооплачуваних віддалених працівників - Software Development - 1186 респондентів з 2461.
2. Далі йдуть індустрії: від 190 до 138 - Fintech, Healthcare, Other:, Internet, Telecomm or Information Services. І тут Fintech та Healthcare активно залучають працівників на віддалену роботу - 190-188.
3. Та інші від 88 до 63. Причому у цьому списку Government - 78, що цікаво.
4. 75-й перцентиль компенсації: $120,596 - це найвищий поріг компенсації.